### PR TITLE
sql+: apd using heap-allocated type

### PR DESCRIPTION
**Pull request by sploiselle**
Here's an initial implementation + set of functions for a new decimal type, which is referred to as `APD` in these commits.

The one big question around the type's semantics that need to be nailed down is: how important is commutativity when using binary addition? We can preserve commutativity in aggregations (i.e. `sum`), but this PR doesn't guarantee it in addition/subtraction. The thinking here is that we can't guarantee it with multiplication/division, so if we have to toss it there, it doesn't make as much sense to maintain it alongside addition. Maybe you'd make a different recommendation, in which case I'd be glad to hear it. Fortunately, if there is a chorus of disapproval, it's easy to change the invariant (just throw `APDFieldOverflow` on inexact sums).

Note that `sum` is implemented and tested in [another branch](https://github.com/sploiselle/materialize/tree/apd-agg-bench-save), but isn't in this PR for difficulty in testing in parallel with the current implementation.

### Features included

- `add_apd` (`+`)
- `sub_apd` (`-`)
- `mul_apd` (`*`)
- `div_apd` (`/`)
- `cast_apd_to_string` (`apd::text`)
- `cast_string_to_apd` (`text::apd`)
- `rescale_apd` (`apd(39,2)::apd(39,1)`)

### CODE REVIEW

**Comment by frankmcsherry** [comment]
I don't have any strong feelings about associativity and commutativity other than that if these would be violated in `sum` then we should throw an error. We happily violate some of these things with `float` types, and .. while I could see folks wanting something better properties I think they are a bonus feature rather than "blocking".

> but isn't in this PR for difficulty in testing in parallel with the current implementation.

Maybe I don't want to know, but .. what is the complication? Is there something in this implementation that makes it hard, or .. what is the path forward?

**Comment by sploiselle** [comment]
@frankmcsherry re:
> what is the complication?

Good question! Testing aggregations is pretty easy when you're...
- decoding values from Avro. This isn't viable because when Avro schemas declare a type as decimal, it can only get routed to one of our data types
- getting values from PG tables. This doesn't work because you can't declare the type "apd" in a table because of symbiosis mode, i.e. the type "apd" DNE in Postgres.

The next easiest thing is to do something like:
- use a CSV source that can take text to the specified type (what I plan to do in the intermediate state). I didn't include that in this PR just for sake of getting something up to allow parallelization of work between Phillip and I, but that change is still incoming.
- test `sum` last, i.e. when we can commit the code that changes the type "decimal" from referring to the current implementation to the new implementation. I mention this only as a kind of worst-case scenario.

**Review by benesch — APPROVED** [review]
LGTM! Not supporting associativity/commutativity in addition/subtraction etc. seems totally fine.

I would like to workshop the APIs in e.g. repr::adt::apd at some point before we ship this, but top priority is just getting this stuff merged for now. Casting to `::apd` is experimental, right? If not it needs to be!

**Comment by philip-stoev** [comment]
Item 1. This does not seem right:

```
materialize=> select '1'::apd / '11'::apd;
ERROR:  value out of range: exceeds maximum precision 39
```

The same computation works with decimals.

**Comment by philip-stoev** [comment]
Item #2. I think this should have errored out with a "division by zero" error, as it does in postgres and as it does with the decimal type:

```
materialize=> select '0'::apd / '0'::apd;
 ?column? 
----------
 NaN
(1 row)
```

**Comment by philip-stoev** [comment]
Item #3. Leading and trailing spaces should be ignored, as they are in postgres and in the mz decimal type:

```
materialize=> select cast('0.123  ' as apd);
ERROR:  invalid input syntax for type apd: invalid decimal syntax: "0.123  "
materialize=> select cast('   0.123' as apd);
ERROR:  invalid input syntax for type apd: invalid decimal syntax: "   0.123"
```

**Comment by philip-stoev** [comment]
Item #4  ```-NaN``` can be constructed, even though this is not legal in Postgres (and in Mz, casting ```-NaN``` as double returns ```NaN```) :

```
materialize=> SELECT '-NaN'::apd;
 ?column? 
----------
 -NaN
(1 row)
```

File: src/repr/src/adt/apd.rs [hunk 1]
```diff
@@ -15,16 +15,34 @@
...
-/// The maximum number of digits expressable in an RDN.
-pub const RDN_MAX_PRECISION: usize = 34;
+/// The maximum number of digits expressable in an APD.
+pub const APD_DATUM_WIDTH: usize = 13;
+pub const APD_DATUM_MAX_PRECISION: usize = APD_DATUM_WIDTH * 3;
+
+lazy_static! {
+    static ref CX_DATUM: Context<Decimal<APD_DATUM_WIDTH>> = {
+        let mut cx = Context::<Decimal<APD_DATUM_WIDTH>>::default();
+        cx.set_max_exponent(isize::try_from(APD_DATUM_WIDTH * 3 - 1).unwrap())
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/repr/src/adt/apd.rs]: ty!

File: src/sql/src/plan/typeconv.rs [hunk 2]
```diff
@@ -349,19 +359,25 @@ fn get_cast(
         }
     }
 
-    if from == to {
-        return Some(Box::new(|expr| expr));
-    }
+    match to {
+        // Do not use equality match if `to` has specified scale
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql/src/plan/typeconv.rs]: Made this less obscure and added a clarifying comment. tl;dr is that I patterned the implementation after decimals originally, but instead moved generating this cast inline instead. lmk if you've a predilection for one form or another.

**Comment by sploiselle** [comment]
@philip-stoev Fixed all of those cases

- Item 1 revealed some changes I needed to make to the API; namely, it's more reasonable to round input to the nearest fifteen-digits-of-precision value rather than prohibiting values with > 15 digits. This makes our input behave more like the results of multiplication and division, rather than having its own unique semantics. I've updated the tests to reflect this more completely.
- Item 2 showed that the underlying library has an odd order of preference for its error states that preferred signalling "invalid operations" over "dividing by zero." Seemed to have fixed that generally.
- Item 3 was an easy fix by adding `trim`
- Item 4 required some special casing to error on that input

**Comment by philip-stoev** [comment]
This produces a negative zero in this branch, but postgres and the Mz DECIMAL type return a "normal" zero.

```
materialize=> SELECT '-1'::APD * '0'::APD ;
 ?column? 
----------
 -0
(1 row)
```

I think those negative zeroes have the potential for confusing users.

**Comment by philip-stoev** [comment]
Item #6. This returns different results with APD and in Postgres, after the 5-th decimal sign:

```
pstoev=# SELECT
pstoev-# '8'::DECIMAL
pstoev-# - '5'::DECIMAL
pstoev-# / '9'::DECIMAL 
pstoev-# + '330627'::DECIMAL 
pstoev-# / '27.35'::DECIMAL 
pstoev-# * '91'::DECIMAL 
pstoev-# * '38.9'::DECIMAL 
pstoev-# * '3'::DECIMAL
pstoev-# * '163'::DECIMAL 
pstoev-# - '.4'::DECIMAL 
pstoev-# + '.4'::DECIMAL 
pstoev-# - '955.3'::DECIMAL
pstoev-# / '-23'::DECIMAL
pstoev-# * '-688'::DECIMAL
pstoev-# / '5' AS DECIMAL ;
             decimal              
----------------------------------
 20925734948.92014842335360174444
(1 row)


materialize=> SELECT
materialize-> '8'::APD
materialize-> - '5'::APD
materialize-> / '9'::APD 
materialize-> + '330627'::APD 
materialize-> / '27.35'::APD 
materialize-> * '91'::APD 
materialize-> * '38.9'::APD 
materialize-> * '3'::APD
materialize-> * '163'::APD 
materialize-> - '.4'::APD 
materialize-> + '.4'::APD 
materialize-> - '955.3'::APD
materialize-> / '-23'::APD
materialize-> * '-688'::APD
materialize-> / '5' AS APD ;
                   apd                    
------------------------------------------
 20925734948.9201490784163067765325137552
```

Apologies beforehand for the ugly query :-(

**Review by philip-stoev — APPROVED** [review]
As disscussed with @benesch , testing will continue after this is merged. I will not be adding any more items to this review.

Randomized testing has been doing OK so far, that is, ```'1'::apd + '1'::apd = 2``` but I need to examine a couple of situations where the APD slightly differs from Postgresql  .

**Pull request merged**